cmd/clusterctl/client/tree/discovery.go

```diff
@@ -93,54 +104,131 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
+					tree.Add(addons, resourceSet, ObjectMetaName("ClusterResourceSet"))
+				}
+			}
+		}
+	}
+}
+
+func addControlPlane(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, controlPlane *unstructured.Unstructured, tree *ObjectTree, options DiscoverOptions) {
+	tree.Add(cluster, controlPlane, ObjectMetaName("ControlPlane"), GroupingObject(true))
+
+	if options.ShowAllResources {
+		// Add control plane infrastructure ref using spec fields guaranteed in contract
+		infrastructureRef, found, err := unstructured.NestedMap(controlPlane.UnstructuredContent(), "spec", "machineTemplate", "infrastructureRef")
+		if err != nil && found {
+			infrastructureObjectRef := &corev1.ObjectReference{
+				Kind:       infrastructureRef["kind"].(string),
+				Namespace:  infrastructureRef["namespace"].(string),
+				Name:       infrastructureRef["name"].(string),
+				APIVersion: infrastructureRef["apiVersion"].(string),
+			}
+			if machineTemplate, err := external.Get(ctx, c, infrastructureObjectRef, cluster.Namespace); err == nil {
+				tree.Add(controlPlane, machineTemplate, ObjectMetaName("MachineInfrastructureTemplate"))
+			}
+		}
+	}
+}
+
+func addMachineDeploymentToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, workers *unstructured.Unstructured, machinesList *clusterv1.MachineList, tree *ObjectTree, options DiscoverOptions, addMachineFunc func(parent client.Object, m *clusterv1.Machine)) error {
 	// Adds worker machines.
 	machinesDeploymentList, err := getMachineDeploymentsInCluster(ctx, c, cluster.Namespace, cluster.Name)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	machineSetList, err := getMachineSetsInCluster(ctx, c, cluster.Namespace, cluster.Name)
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	for i := range machinesDeploymentList.Items {
 		md := &machinesDeploymentList.Items[i]
 		addOpts := make([]AddObjectOption, 0)
-		if !options.ShowMachineSets {
+		if !options.ShowMachineSets && !options.ShowAllResources {
 			addOpts = append(addOpts, GroupingObject(true))
 		}
 		tree.Add(workers, md, addOpts...)
 
+		if options.ShowAllResources {
+			if bootstrapTemplate, err := external.Get(ctx, c, md.Spec.Template.Spec.Bootstrap.ConfigRef, cluster.Namespace); err == nil {
+				tree.Add(md, bootstrapTemplate, ObjectMetaName("BootstrapConfigTemplate"))
+			}
+
+			if machineTemplate, err := external.Get(ctx, c, &md.Spec.Template.Spec.InfrastructureRef, cluster.Namespace); err == nil {
+				tree.Add(md, machineTemplate, ObjectMetaName("MachineInfrastructureTemplate"))
+			}
+		}
+
 		machineSets := selectMachinesSetsControlledBy(machineSetList, md)
 		for i := range machineSets {
 			ms := machineSets[i]
 
 			var parent client.Object = md
-			if options.ShowMachineSets {
-				tree.Add(md, ms, GroupingObject(true))
+			if options.ShowMachineSets || options.ShowAllResources {
```

Comment:
Also, in this case, I'm personally -1 to mix up the effects of --allresources and --machineset flag, this can be confusing for the users.